sympy/core/function.py

```diff
@@ -320,7 +324,7 @@ def __new__(cls, *args, **options):
             #  - AppliedUndef with no nargs like Function('f')(1).nargs
             nargs = obj._nargs  # note the underscore here
         # convert to FiniteSet
-        obj.nargs = FiniteSet(*nargs) if nargs else Naturals0()
+        obj.nargs = _FiniteSetCache(*nargs) if nargs else S.Naturals0
```

Comment:
Is there a reason this has to be set in the constructor? Otherwise we could just add @cacheit to the nargs property definition in FunctionClass.

Reply:
It was there, so I did not move it